File: Code/07_Pipes_Sinais_Alarmes/Ex1.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <string.h>

int main()
{
	int pid;	// PID do processo filho
	int fd[2];	// Descritores do pipe
	char mensagem[30];
	pipe(fd); // Cria o pipe
	pid = fork(); // Cria o processo
	if(pid == 0) // Codigo do filho
	{
		printf("Filho vai ler o pipe\n");
		if(read(fd[0], mensagem, 30) < 0)
			printf("Erro na leitura do pipe\n");
		else
			printf("Filho leu: %s\n", mensagem);
	}
	else // Codigo do pai
	{
		strcpy(mensagem, "HELLO PIPE");
		printf("Pai vai escrever no pipe\n");
		if (write(fd[1], mensagem, 30) < 0)
			printf("Erro na escrita do pipe\n");
		printf("Pai terminou de escrever no pipe\n");
	}
	return 0;
}

File: 2_Programação_sistemas_operacionais/2.4_Pipes_Sinais_Alarmes/Python/Ex1.py
import os

fd = os.pipe()
pid_filho = os.fork()
if pid_filho == 0:
	print("Filho vai ler o pipe")
	mensagem = os.read(fd[0], 30)
	if(len(mensagem)==0):
		print("Erro na leitura do pipe")
	else:
		print("Filho leu: " +  mensagem.decode("utf-8"))
else:
	mensagem = "HELLO PIPE"
	# Vamos garantir o envio de 30 bytes no pipe
	# inserindo 20 caracteres '\0'
	mensagem += "\0"*(30-len(mensagem)) 
	
	print("Pai vai escrever no pipe")
	if os.write(fd[1], str.encode(mensagem)) < 0:
		print("Erro na escrita do pipe")
	print("Pai terminou de escrever no pipe")